Refactor/orders fase2 cleanarch

mnunes182 opened the pull request.

mnunes182 committed:
implementa endpoints de checkout e atualização de status na Clean Architecture

mnunes182 committed:
fix: Corrige persistence base

mnunes182 committed:
fix: Database

mnunes182 committed:
fix: import de ProductOrderDTO no CheckoutOrderDTO

mnunes182 committed:
fix: corrigido field dto

mnunes182 committed:
fix: Repositories

mnunes182 committed:
fix(orders): ajustes em controllers, rotas, use cases e testes para estabilizar execução da feature

mnunes182 committed:
fix: ajustes de lint com Ruff e correção de imports nos testes unitários

mnunes182 committed:
fix: quebra de linha em get_by_status para conformidade

mnunes182 committed:
fix: ajustes nos testes de ListProductsService e PaymentService

mnunes182 committed:
fix: corrigidos os testes de listagem de produtos e pagamento

mnunes182 committed:
Merge da branch refactor/clean-architecture para manter atualizações

PauloVitorSchio committed:
Merge branch 'refactor/clean-architecture' into refactor/orders-fase2-cleanarch

PauloVitorSchio commented on app/adapters/controllers/orders/update_order_status_controller.py:
```diff
@@ -0,0 +1,13 @@
+from app.infrastructure.dependencies.order_dependencies import get_update_order_status_use_case
+
+
+def update_status_endpoint(
```
Em vez de chamar de endpoint, chame de controller para manter o padrão de nomenclatura.
`update_status_endpoint` -> `update_order_status_controller`
seria um nome mais descritivo e condizente.

PauloVitorSchio commented on app/adapters/controllers/orders/update_order_status_controller.py:
```diff
@@ -0,0 +1,13 @@
+def update_status_endpoint(
+    order_id: int,
+    dto: UpdateOrderStatusDTO,
+    use_case: UpdateOrderStatus = Depends(get_update_order_status_use_case)
```
A controller deve receber o use_case já instanciado através de injeção de dependência. Lembra que na Clean Architecture, as dependências devem sempre apontar para dentro. Nesse caso, a camada de adapters está importando o Depends do FastAPI e o get_update_order_status_use_case da camada de cima.

Em vez disso, a própria camada de cima já vai instanciar tudo e passar a implementação concreta para a controller. Se tiver dúvida, dá uma olhada em como eu fiz nos outros endpoints.

PauloVitorSchio commented on app/application/use_cases/orders/update_order_status.py:
```diff
@@ -0,0 +1,17 @@
+from app.adapters.gateways.order_repository_gateway import OrderRepositoryGateway
```
Aqui, mesma coisa que eu comentei em outro arquivo, o seu use_case fica em uma camada abaixo dos adapters, então ele não pode importar nada de lá. O que você faz, nesse caso, é definir uma interface na sua camada application, e o OrderRepositoryGateway vai depender dela, implementando-a.

PauloVitorSchio commented on app/application/use_cases/orders/update_order_status.py:
```diff
@@ -0,0 +1,17 @@
+            OrderStatus.FINISHED,
+        ]:
+            raise ValueError("Status inválido.")
+        return self.repository.update(order_id, {"status": status})
```
Aqui acho que dá para melhorar. Perceba que você está atualizando o atributo do objeto direto no repository. Porém, o mais ideal seria que a própria entidade fizesse isso, pois é ela quem encapsula as lógicas de validação, e o repository só teria a função de persistir os dados.

Sugiro algo próximo a isso aqui:
```py
class UpdateOrderStatus:
    def __init__(self, repository: OrderRepositoryGateway):
        self.repository = repository

    def execute(self, order_id: int, new_status: str) -> Order:
        order = self.repository.get_by_id(order_id)
        order.update_status(new_status)
        return self.repository.update(order_id, order)
```
Assim, as regras de negócio ficam encapsuladas na entidade. Aí no update_status você pode validar se o status é válido.

PauloVitorSchio commented on app/config/database.py:
```diff
@@ -1,8 +1,11 @@
 from sqlalchemy import create_engine
 from sqlalchemy.orm import sessionmaker
 
+load_dotenv()
```
Não precisa disso aqui, porque o próprio container já está populando esses valores via variáveis de ambiente.

PauloVitorSchio commented on app/infrastructure/dependencies/__init__.py:
Esse arquivo acredito que veio junto por engano.

PauloVitorSchio commented on app/infrastructure/persistence/repositories/order_repository.py:
Esses métodos novos todos que você criou aqui no repository, você precisaria adicionar também no gateway.

PauloVitorSchio reviewed (changes requested):
Comentários mais gerais:

Cara, sinceramente, eu acho que a parte do checkout, como eu até comentei no grupo, é basicamente o use_case de criar um pedido, que já tava até pronto. Se você ler lá no doc do tech-challenge fase 2, vai ver que os requisitos são idênticos.

Eu vi também que você não implementou o caso de uso de listar os pedidos que ordenava por status, excluía os finalizados, etc.

Outra coisa, quando você fez o merge, ficaram alguns conflitos e você acabou subindo eles junto. Eu dei um pull local na sua branch e resolvi eles, 

Tá bem codado o que você fez, os comentários que eu fiz são referentes a uns detalhes conceituais da Clean Arch.

PauloVitorSchio committed:
fix: Alteração no nome da classe

mnunes182 committed:
refactor(order): integra update_order_status use case, adiciona métodos no gateway e ordenação por status no repository

mnunes182 committed:
feat(order): implement use case for updating order status and map enum filtering
- Add UpdateOrderStatus use case
- Extend OrderRepositoryGateway with update_status method
- Map OrderStatus enum in get_all by status
- Resolve merge conflicts from refactor/orders-fase2-cleanarch

PauloVitorSchio committed:
Merge branch 'main' of https://github.com/tech-challenge-pos-fiap/tech-challenge into refactor/orders-fase2-cleanarch

PauloVitorSchio committed:
fix: Ruff checks

PauloVitorSchio committed:
feat: Implementa atualização de status do pedido

PauloVitorSchio committed:
fix: Configuração de testes

PauloVitorSchio committed:
fix: Remove testes obsoletos temporariamente
